### PR TITLE
Add 'exact match' badge

### PR DESCRIPTION
**Pull request by etvorun**
If a search query matches package ID (sans spaces and case sensitivity) then we display distinct "exact match" badge next to package name. Using existing green themed colors.

Addresses https://github.com/NuGet/NuGetGallery/issues/7463

ExactMatchLight

ExactMatchDark

### CODE REVIEW

File: src/NuGetGallery/Helpers/ViewModelExtensions/ListPackageItemViewModelFactory.cs [hunk 1]
```diff
@@ -48,6 +48,7 @@ private ListPackageItemViewModel SetupInternal(ListPackageItemViewModel viewMode
             viewModel.IsVerified = package.PackageRegistration?.IsVerified;
             viewModel.IsDeprecated = package.Deprecations?.Count > 0;
             viewModel.IsVulnerable = package.VulnerablePackageRanges?.Count > 0;
+            viewModel.IsExactMatch = string.Equals(package.Id, query, StringComparison.OrdinalIgnoreCase);
```
Comment (joelverhagen) [inline src/NuGetGallery/Helpers/ViewModelExtensions/ListPackageItemViewModelFactory.cs:51]: Maybe trim whitespace on the query before comparing, just in case

**Review by joelverhagen — DISMISSED** [review]

File: src/NuGetGallery/Helpers/ViewModelExtensions/ListPackageItemViewModelFactory.cs (hunk 1 above)
Comment (joelverhagen) [reply src/NuGetGallery/Helpers/ViewModelExtensions/ListPackageItemViewModelFactory.cs:51]: Trim might be done elsewhere, worth checking.

Comment (etvorun) [reply src/NuGetGallery/Helpers/ViewModelExtensions/ListPackageItemViewModelFactory.cs:51]: I wonder how much value it would add? There are already multiple places and tests ensuring that we pass around trimmed search query, e.g. https://github.com/NuGet/NuGetGallery/blob/77cba4f2ac7b0d9e669c999d0ca3aff4cae5dfd2/src/NuGetGallery/Controllers/PackagesController.cs#L1282C13-L1282C44 
https://github.com/NuGet/NuGetGallery/blob/77cba4f2ac7b0d9e669c999d0ca3aff4cae5dfd2/tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs#L5040

Comment (joelverhagen) [reply src/NuGetGallery/Helpers/ViewModelExtensions/ListPackageItemViewModelFactory.cs:51]: I'm thinking if someone pastes a package ID in search that has leading or trailing whitespace, the exact match badge should still work.

Comment (joelverhagen) [reply src/NuGetGallery/Helpers/ViewModelExtensions/ListPackageItemViewModelFactory.cs:51]: Just something worth testing.

File: src/Bootstrap/less/theme/page-list-packages.less [hunk 2]
```diff
@@ -285,6 +285,17 @@
     }
   }
 
+  .exact-match-badge {
+    padding: 1px 4px;
+    border-radius: 4px;
+    display: inline-block;
+    border-radius: 4px;
```
Comment (agr) [inline src/Bootstrap/less/theme/page-list-packages.less]: `border-radius` is specified twice.

Comment (agr) [reply src/Bootstrap/less/theme/page-list-packages.less]: Do we have a variable for "badge border radius", btw? It's not the only badge that we use. If not, might be worth adding one.

File: tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs [hunk 3]
```diff
@@ -5053,6 +5053,41 @@ public async Task TrimsSearchTerm()
...
+                searchService
+                    .Setup(s => s.Search(It.IsAny<SearchFilter>()))
+                    .ReturnsAsync(new SearchResults(1, DateTime.UtcNow, packages.AsQueryable()));
+
+                var result = await controller.ListPackages(new PackageListSearchViewModel { Q = query });
+
+                var model = ResultAssert.IsView<PackageListViewModel>(result);
+                var matchedItems = model.Items.Where(p => p.IsExactMatch).ToList();
+                Assert.Single(matchedItems);
+                Assert.Equal("test", matchedItems[0].Id);
```
Comment (agr) [inline tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs]: The three lines above can be replaced with something like:

```C#
Assert.Equal("test", Assert.Single(model, p => p.IsExactMatch).Id);
```

File: tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs [hunk 4]
```diff
@@ -5053,6 +5053,41 @@ public async Task TrimsSearchTerm()
                 Assert.Equal("test", model.SearchTerm);
             }
 
+            [Theory]
+            [InlineData("test")]
+            [InlineData("TEST")]
```
Comment (agr) [inline tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs]: More test cases? Might need some negative test cases, too (in a separate test method, likely).
```
"test "
" test"
"test+"
```

File: src/NuGetGallery/Helpers/ViewModelExtensions/ListPackageItemViewModelFactory.cs (hunk 1 above)
Comment (etvorun) [reply src/NuGetGallery/Helpers/ViewModelExtensions/ListPackageItemViewModelFactory.cs:51]: Tests added

File: tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs (hunk 3 above)
Comment (etvorun) [reply tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs]: Is such level of density required by our coding standards? I tend to avoid performing multiple validations on one line. For better debuggability.

File: src/Bootstrap/less/theme/page-list-packages.less (hunk 2 above)
Comment (etvorun) [reply src/Bootstrap/less/theme/page-list-packages.less]: Would it make sense to normalize border-radius and other values in a separate PR? There are lots of repeating numbers for common properties in *.less files.

File: tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs (hunk 4 above)
Comment (etvorun) [reply tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs]: Added

File: tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs [hunk 5]
```diff
@@ -5053,6 +5053,53 @@ public async Task TrimsSearchTerm()
...
+                searchService
+                    .Setup(s => s.Search(It.IsAny<SearchFilter>()))
+                    .ReturnsAsync(new SearchResults(1, DateTime.UtcNow, packages.AsQueryable()));
+
+                var result = await controller.ListPackages(new PackageListSearchViewModel { Q = query });
+
+                var model = ResultAssert.IsView<PackageListViewModel>(result);
+                var matchedItems = model.Items.Where(p => p.IsExactMatch).ToList();
+                if (expectedMatch)
+                {   Assert.Single(matchedItems);
```
Comment (agr) [inline tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs]: Missing line break.

Comment (etvorun) [reply tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs]: Good catch :)

**Review by agr — APPROVED** [review]
